### ISSUE TITLE
Only grant `connect,accept` permissions to `transport-netty4`

### ISSUE THREAD

**Issue by s1monw** [elastic/elasticsearch#22116]
Today we still grant a quite scary permission to [core](https://github.com/elastic/elasticsearch/blob/master/core/src/main/resources/org/elasticsearch/bootstrap/security.policy#L58):

```
  // Allow connecting to the internet anywhere
  permission java.net.SocketPermission "*", "accept,connect,resolve";
```

But since we now have netty4 moved to a module we can potentially move this to into the modules [security policy](https://github.com/elastic/elasticsearch/blob/master/modules/transport-netty4/src/main/plugin-metadata/plugin-security.policy). Yet, there are a bunch of things that needs fixing until we can do that:

- [x] since we use [MockTcpTransport](https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java) from our test framework this needs to have the same permissions granted. Yet, if we just go ahead and grant `accept,connect` to the test-framework we might run into trouble since our tests will just inherit that permission ie. if unit and pseudo integ-tests are run since we don't grant this to a codebase. We might want to add some kind of `MockSocket` project just like [SecureMock](https://github.com/elastic/securemock) that we can grant this permission to and where we can depend on for testing. 
- [x] netty-4 still has issues with missing `doPrivileged` blocks that needs fixing
- [x] move URLRepository somewhere else since it uses _connect_ and core shouldn't establish any kind of connection. (this can be a second step, we can first start removing `accept` from the list.
- [x] some other plugins like ec2 / gce etc. might need extra permission to connect to their endpoints which needs manual testing

here is an example of a missing `doPrivileged` block  ie here: 
``` 
> Throwable #1: java.security.AccessControlException: access denied ("java.net.SocketPermission" "[fe80:0:0:0:0:0:0:1%1]:52661" "connect,resolve")
   > 	at __randomizedtesting.SeedInfo.seed([8FDA867CA1C20E0D:47AB82401EC3F2C5]:0)
   > 	at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
   > 	at java.security.AccessController.checkPermission(AccessController.java:884)
   > 	at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
   > 	at java.lang.SecurityManager.checkConnect(SecurityManager.java:1051)
   > 	at sun.nio.ch.SocketChannelImpl.connect(SocketChannelImpl.java:625)
   > 	at io.netty.channel.socket.nio.NioSocketChannel.doConnect(NioSocketChannel.java:331)
   > 	at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.connect(AbstractNioChannel.java:254)
   > 	at io.netty.channel.DefaultChannelPipeline$HeadContext.connect(DefaultChannelPipeline.java:1266)
   > 	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:556)
   > 	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:541)
   > 	at io.netty.channel.ChannelOutboundHandlerAdapter.connect(ChannelOutboundHandlerAdapter.java:47)
   > 	at io.netty.channel.AbstractChannelHandlerContext.invokeConnect(AbstractChannelHandlerContext.java:556)
   > 	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:541)
   > 	at io.netty.channel.AbstractChannelHandlerContext.connect(AbstractChannelHandlerContext.java:523)
   > 	at io.netty.channel.DefaultChannelPipeline.connect(DefaultChannelPipeline.java:985)
   > 	at io.netty.channel.AbstractChannel.connect(AbstractChannel.java:255)
   > 	at io.netty.bootstrap.Bootstrap$3.run(Bootstrap.java:252)
   > 	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
   > 	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:418)
   > 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:454)
   > 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:873)
   > 	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
   > 	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jasontedor** [comment]
+1

**Comment by tbrooks8** [comment]
I've been able to make the modifications to netty necessary to move the permissions to the transport plugin and get the netty related and core tests passing (with creation of a mocksocket jar).

I've had some issues with this related to azure plugin. That plugin uses com.sun.net.httpserver.HttpsServer for running some tests. it appears to me that HttpsServer is not calling doPriviledge when accepting connections.

```
if(!ServerImpl.this.terminating) { 
    SocketChannel var5 = ServerImpl.this.schan.accept();
     if(ServerConfig.noDelay()) { 
        var5.socket().setTcpNoDelay(true); 
    }  
    if(var5 != null) {
         var5.configureBlocking(false);
         SelectionKey var6 = var5.register(ServerImpl.this.selector, 1);
         var7 = new HttpConnection();
         var7.selectionKey = var6; 
        var7.setChannel(var5); 
        var6.attach(var7);
         ServerImpl.this.requestStarted(var7); 
        ServerImpl.this.allConnections.add(var7);
     } 
}
```

**Comment by s1monw** [comment]
@tbrooks8  amazing progress, would you mind linking the netty PRs / issues with this issue so we can track them. Also for all core changes can you make sure you open smallish PRs if possible?

> I've had some issues with the azure plugin. That plugin uses com.sun.net.httpserver.HttpsServer for running some tests. it appears to me that HttpsServer is not calling doPriviledge when accepting connections.

that is super annoying. I am not 100% sure what we can do about this, I think first of all we gotta fix it in the JDK such that down the road we can remove whatever hack we add here.

**Comment by tbrooks8** [comment]
I push the work that I have completed on the mocksocket here:

https://github.com/elastic/mocksocket

I have opened a Netty PR (netty/netty#6146)

I will start making PRs to ES as I split up the work.

**Comment by s1monw** [comment]
cool stuff - can you make the `https://github.com/elastic/mocksocket` repo public I don't think think there is anything private about this! :)

**Comment by tbrooks8** [comment]
I have opened a PR to integrate mocksocket into ElasticSearch.

#22287

**Comment by tbrooks8** [comment]
Based on the discussion in the status meeting today, I investigated applying permissions based on files opposed to a separate jar.

So I migrated `accept` out of the core security policy. And placed the following the in test-framework.policy:

```
grant codeBase "file:/path/to/elastic/elasticsearch/test/framework/build-idea/classes/test/" {
  permission java.net.SocketPermission "*", "accept,resolve,connect";
 };

grant codeBase "file:/path/to/elastic/elasticsearch/test/framework/build-idea/classes/main/" {
   permission java.net.SocketPermission "*", "accept,resolve,connect";
 };
```

With that change, any tests involving the `MockTCPTransport` passed. (Although the netty transport tests failed as they were not given permission.)

This approach seems to provide a way forward. Although my approach is certainly naive and hardcoded.

I'm not immediately clear on what approach to take in order to make this solution robust. I see that we currently [parse the classpath and build up some system properties](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/bootstrap/Security.java#L194). Currently this works pretty well for jars. However, in the case of our code it is constantly rewriting the `codebase.main` and `codebase.test` system properties.

So when you run the "core" tests the following codebases are parsed:

```
file:/Users/timbrooks/elastic/elasticsearch/core/build-idea/classes/test/
file:/Users/timbrooks/elastic/elasticsearch/core/build-idea/classes/main/
file:/Users/timbrooks/elastic/elasticsearch/test/framework/build-idea/classes/main/
file:/Users/timbrooks/elastic/elasticsearch/client/rest/build-idea/classes/main/
```

And for discovery-azure-classic plugin tests:

```
file:/Users/timbrooks/elastic/elasticsearch/plugins/discovery-azure-classic/build-idea/classes/test/
file:/Users/timbrooks/elastic/elasticsearch/plugins/discovery-azure-classic/build-idea/classes/main/
file:/Users/timbrooks/elastic/elasticsearch/plugins/discovery-azure-classic/build-idea/generated-resources/
file:/Users/timbrooks/elastic/elasticsearch/core/build-idea/classes/main/
file:/Users/timbrooks/elastic/elasticsearch/test/framework/build-idea/classes/main/
file:/Users/timbrooks/elastic/elasticsearch/client/rest/build-idea/classes/main/
```

Our method of parsing the "shortName" means that the property will be `codebase.test`, `codebase.main`, or `codebase.generated-source`. And whichever one is parsed last "wins".

I could work on implementing proper handling of this. Such as parsing these into:
- `codebase.core.main`
- `codebase.core.test`
- `codebase.test.framework.main`
- `codebase.test.framework.test`
- `codebase.plugins.discovery-azure-classic.main`
- `codebase.plugins.discovery-azure-classic.test`

Those system properties could be accessed in security policy files to give certain packages permissions. Or you should be able to narrow it down to specific classes. Such as:

```
grant codeBase "${codebase.test.framework.main}/org/elasticsearch/transport/MockTcpTransport.class" {
  // Permissions
}
```

Maybe this is not the easiest way forward? It's entirely possible that I missing a better way of doing this.

**Comment by s1monw** [comment]
++ this sounds like a plan @tbrooks8 

> Also - I'm not sure what implications those changes would have for the distribution (when all the packages / modules are different jars).

I think that will be fine since we don't use this in distribution. Our tests will catch if something is wrong since they use the real zip / tar.gz

**Comment by tbrooks8** [comment]
I've spiked on this [here](https://github.com/tbrooks8/elasticsearch/commit/a100ea87dd1b0eed85eb078fdc694ad1156b4815#diff-3be67d43cef62609fb9f591011401559R198).

I've been learning more about the `SecurityManager` and it seems that creating a `new Thread` inherits the permissions from where the thread was created. So giving "discovery-azure-classic" `SocketPermission` and wrapping https `start` in `doPrivileged {}` with proper permissions for  discovery-azure-classic works.

```
AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
            httpsServer.start();
            httpServer.start();
            return null;
        });
```
combined with:

```
grant codeBase "${codebase.elasticsearch.plugins.discovery-azure-classic.test}" {
  permission java.net.SocketPermission "*", "accept,connect,resolve";
};

grant codeBase "${codebase.elasticsearch.plugins.discovery-ec2.test}" {
  permission java.net.SocketPermission "*", "accept,connect,resolve";
};

grant codeBase "${codebase.elasticsearch.plugins.discovery-gce.test}" {
  permission java.net.SocketPermission "*", "accept,connect,resolve";
};
```

allows all of the tests for those packages to pass.

There are a few things about this that are kind of ugly. If you run the tests in an IDE (like IntelliJ) the main classes remain in directories. If you run the tests using Gradle the main classes are placed in an JAR. This means that for test "framework" you must use two permissions for `MockTCPTransport` to work.

```
grant codeBase "${codebase.elasticsearch.test.framework.main}" {
  permission java.net.SocketPermission "*", "accept,connect,resolve";
};

grant codeBase "${codebase.framework-6.0.0-alpha1-SNAPSHOT.jar}" {
  permission java.net.SocketPermission "*", "accept,connect,resolve";
};
```

I think that tomorrow I should explore moving the starting of the HTTPS and HTTP servers to the MockSocket JAR. And if that works then we should still be able to isolate all `SocketPermission` to that JAR for test packages. And avoid scattering `SocketPermission` to many different packages (framework main, framework test, ec2 test, azure test, etc).

**Comment by tbrooks8** [comment]
Updates:

1. I updated #22287 with changes to mocksocket that include wrapping the http/https servers.
2. I pushed mocksocket to MavenCentral.
3. I worked on integrating the netty snapshot based on my PR on this [branch](https://github.com/tbrooks8/elasticsearch/tree/integrate_netty_updates). With the netty snapshot + mocksocket PR, when I move the `accept` permissions out of core and to just netty/mocksocket the elasticsearch test suite passes.

**Comment by tbrooks8** [comment]
I've started exploring what is involved in removing `connect` and `resolve` from core.

Here are some preliminary findings:

1. `NetworkUtils` makes the following call that requires connect:
- `NetworkInterface.getInetAddresses()`
2. `MacAddressProvider` makes the following call that requires connect:
- `NetworkInterface.getHardwareAddress()`
3. `IfConfig` uses these two calls when logging.

So maybe these classes will need to be moved to a dependency? Or something else?

As Simon pointed out various plugins (ec2, etc) will need permissions granted to a number of their dependencies (EC2Client, etc). I have not figured out complete lists of these yet.

**Comment by s1monw** [comment]
`NetworkInterface.getInetAddresses()` & `NetworkInterface.getHardwareAddress()` supports the `getNetworkInformation` permission I think we should grant this to get rid of one of the problems.

For `IfConfig.logIfNecessary();` I wonder if we can move it into bootstrap instead then we can log it before we initialize the security manager? @jasontedor WDYT?

> As Simon pointed out various plugins (ec2, etc) will need permissions granted to a number of their dependencies (EC2Client, etc). I have not figured out complete lists of these yet.

yeah I think all the repositories need resolve an connect granted.

**Comment by tbrooks8** [comment]
@simonw I do not know if moving `IfConfig.logIfNecessary();` to bootstrap was decided on as the correct option, but I opened a PR with that work.

**Comment by jasontedor** [comment]
> For `IfConfig.logIfNecessary();` I wonder if we can move it into bootstrap instead then we can log it before we initialize the security manager? @jasontedor WDYT?

+1

**Comment by tbrooks8** [comment]
As a quick update - it does appear that if I grant permissions to a plugin, that does successfully grant permissions to all of the dependency jars. So after I add the `doPrivilege` blocks the plugins (ec2, gce, etc) should work.

However, I identified a few other issues. In two places in core we currently need connect/resolve. In `NetworkService` ([here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/network/InetAddresses.java#L352)) and `TcpTransport` ([here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/transport/TcpTransport.java#L822)).

Thoughts?

**Comment by s1monw** [comment]
> However, I identified a few other issues. In two places in core we currently need connect/resolve. In NetworkService (here) and TcpTransport (here).

hmm I am starting to think maybe removing `connect, resolve` is too ambiguous. I don't think it's worth adding a specific jar we depend on to grant this permission so I wonder if we should stick with `accept` for now? @jasontedor  WDYT?

**Comment by jasontedor** [comment]
> However, I identified a few other issues. In two places in core we currently need connect/resolve. In `NetworkService` ([here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/network/InetAddresses.java#L352)) and `TcpTransport` ([here](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/transport/TcpTransport.java#L822)).

@tbrooks8 I think your first link here is not right? At least, I don't think what you link to needs permissions.

**Comment by jasontedor** [comment]
I'm not sure @s1monw, I wonder if we can just add logic to `SecureSM` to handle these permissions the way that we want to (similar to how we handle `System#exit`). We already have a separate jar for this. Thoughts?

**Comment by tbrooks8** [comment]
@jasontedor I meant https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/network/NetworkService.java#L266

**Comment by jasontedor** [comment]
@tbrooks8 Okay, thanks for clarifying. That will require `resolve`.

**Comment by jasontedor** [comment]
Aren't both of the places that you link to only requiring resolve? I'm okay with `resolve`, it's `connect` that might be worth jumping through hoops for. Can you confirm @tbrooks8?

**Comment by tbrooks8** [comment]
It looks to me that InetAddress.getAllByName(host) will only require resolve.

**Comment by tbrooks8** [comment]
Inside `getAllByName0`:

```
 if (check) {
    SecurityManager security = System.getSecurityManager();
    if (security != null) {
        security.checkConnect(host, -1);
    }
}
```

Which calls:

```
public void checkConnect(String host, int port) {
        if (host == null) {
            throw new NullPointerException("host can't be null");
        }
        if (!host.startsWith("[") && host.indexOf(':') != -1) {
            host = "[" + host + "]";
        }
        if (port == -1) {
            checkPermission(new SocketPermission(host,
                SecurityConstants.SOCKET_RESOLVE_ACTION));
        } else {
            checkPermission(new SocketPermission(host+":"+port,
                SecurityConstants.SOCKET_CONNECT_ACTION));
        }
    }
```

**Comment by tbrooks8** [comment]
I explored Jason's thought quite a bit on Friday. It appears that moving connect to the plugins (and netty) and leaving resolve in core should be a viable strategy. 

I'm working on a PR to put `doPrivileged` blocks in the ec2, gce, and azure plugins.

**Comment by s1monw** [comment]
cool thx for the update

**Comment by tbrooks8** [comment]
I opened #22534 to wrap the `connect` operations in plugins.

**Comment by tbrooks8** [comment]
- I opened PR #22572 that should remove a netty release as a blocker for migrating accept and connect.
- I have investigated URLRepository and it does not appear to me that class makes socket connections (some of its plugin subclasses do, but I'm dealing with that in the plugin changes). Let me know if I am incorrect about that. 
- I opened PR #22603 to wrap connect operations in the rest client in doPrivileged. This is necessary for modules such as reindex.
- @jasontedor has suggested that we add a permission like SpecialPermission("connect") for socket specific permissions in ES. Maybe this is a new issue or task?
- @s1monw suggested in PR #22534 that we leverage forbidden APIs to detect improper connect calls. Maybe this is another task for this issue?

**Comment by s1monw** [comment]
>  have investigated URLRepository and it does not appear to me that class makes socket connections (some of its plugin subclasses do, but I'm dealing with that in the plugin changes). Let me know if I am incorrect about that.

this might happen [here]?(https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/blobstore/url/URLBlobContainer.java#L105)

**Comment by tbrooks8** [comment]
Yep. I agree that looks like a connect. Thanks. I must of missed that.

**Comment by tbrooks8** [comment]
I've looked into what is involved in moving `URLRepository` to a module. It looks pretty straightforward to me to move just `URLRepository`.

@s1monw suggested that I move both of the core `Repository` classes (`URLRepository` and `FsRepository`). Unfortunately moving `FsRepository` looks more complicated. `MockRepository` extends `FsRepository`. The following core tests use `MockRepository`:

- `BlobStoreFormatIT`
- `DedicatedClusterSnapshotRestoreIT`
- `MinThreadsSnapshotRestoreIT`
- `RepositoriesIT`
- `RepositoryUpgradabilityIT`
- `RestoreBackwardsCompatIT`
- `SharedClusterSnapshotRestoreIT`

And the functionality seems to depend on `FsBlobStore` and `FsBlobContainer` (so the super class cannot simply be removed). 

I guess there are a couple of options:
- Those tests be moved to the new repositories module
- `MockRepository` be modified to stand alone without the `FSRepository` components. (Also maybe `MockRepository` should be moved to `test.framework` opposed to `core-tests`?)
- Those tests be moved to an external qa package that depends on both `core` and `repositories`.
- Some combination of the above? 

Also - should this be completed in multiple PRs? Like an initial PR that moves `URLRepository` to modules that unblocks SocketPermission work. And then another issue/PR to tackle the moving `FSRepository`?

**Comment by s1monw** [comment]
ok I think we start with URLRepository and take it from there? I didn't see how much FSRepo was involved here so sorry for pushing that direction. The cleanups would be nice but lets focus on the connect/accept permissions for now.

**Comment by tbrooks8** [comment]
I opened the PR #22622  that should removed `connect`.

**Comment by tbrooks8** [comment]
Once #23057 is merged in the last checkbox will be complete.

As an additional note #4815 marked `connect` apis as forbidden as at attempt to warn developers not to use those apis or to provide appropriate permissions.

 @s1monw Once the last PR is merged in is this issue complete?

**Comment by s1monw** [comment]
@tbrooks8 I think we are done here... congrats!

**Comment by tbrooks8** [comment]
Closing as everything has been merged.
